### PR TITLE
Fixes and optimizes numberOfDigits. Improve log2, log10. Adds epsilon to almostEquals and almostZero.

### PR DESCRIPTION
**Pull request by reosfire**
Currently number of digits works incorrectly (it computes radix.toString(radix).len instead of this.toString(radix).len) and also slow. Also I made some optimizations in Int.log functions. 
And added optional epsilon parameter to the almostEquals function

### CODE REVIEW

File: korlibs-math-core/src/korlibs/math/Math.kt [hunk 1]
```diff
@@ -124,10 +134,27 @@ fun max(a: Double, b: Double, c: Double, d: Double, e: Double) = max(max(max(max
...
+fun Long.numberOfDigits(radix: Int = 10): Int {
+    if (this == 0L) return 1
+    var n = this
+    var digits = 0
+    while (n != 0L) {
+        n /= radix
+        digits++
+    }
+    return digits
+}
```
Comment (soywiz) [inline korlibs-math-core/src/korlibs/math/Math.kt:162]: `log` should be faster for this in most modern processors. Tested on x64 and ARM64. Here's a benchmark:

https://jsbenchmark.com/#eyJjYXNlcyI6W3siaWQiOiJ6bExvXzhjUkdSZGxFcjM0UHpEV2IiLCJjb2RlIjoiREFUQS5udW1iZXJPZkRpZ2l0czEoTWF0aC5yYW5kb20oKSAqIDEwMDAwMDAwKSIsIm5hbWUiOiJEaXZpc2lvbiIsImRlcGVuZGVuY2llcyI6W119LHsiaWQiOiJEMmhfZ0tRSkVfWWhLdUV0SVFLLVciLCJjb2RlIjoiREFUQS5udW1iZXJPZkRpZ2l0czIoTWF0aC5yYW5kb20oKSAqIDEwMDAwMDAwKSIsIm5hbWUiOiJMb2ciLCJkZXBlbmRlbmNpZXMiOltdfV0sImNvbmZpZyI6eyJuYW1lIjoiQmFzaWMgZXhhbXBsZSIsInBhcmFsbGVsIjp0cnVlLCJnbG9iYWxUZXN0Q29uZmlnIjp7ImRlcGVuZGVuY2llcyI6W119LCJkYXRhQ29kZSI6ImZ1bmN0aW9uIG51bWJlck9mRGlnaXRzMSh0aGF0LCByYWRpeCA9IDEwKSB7XG4gICAgaWYgKHRoYXQgPT0gMCkgcmV0dXJuIDFcbiAgICB2YXIgbiA9IHRoYXRcbiAgICB2YXIgZGlnaXRzID0gMFxuICAgIHdoaWxlIChuICE9IDApIHtcbiAgICAgICAgbiA9IChuIC8gcmFkaXgpIHwgMFxuICAgICAgICBkaWdpdHMrK1xuICAgIH1cbiAgICByZXR1cm4gZGlnaXRzXG59XG5cbmZ1bmN0aW9uIG51bWJlck9mRGlnaXRzMih0aGF0LCByYWRpeCA9IDEwKSB7XG4gICAgcmV0dXJuIChNYXRoLmxvZyh0aGF0LCByYWRpeCl8MCkgKyAxXG59XG5cbnJldHVybiB7IG51bWJlck9mRGlnaXRzMSwgbnVtYmVyT2ZEaWdpdHMyIH0ifX0

image

File: korlibs-math-core/test/korlibs/math/MathTest.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
+package korlibs.math
+
+import kotlin.test.*
+
+class MathTest {
+    @Test
+    fun testIntNumberOfDigits() {
+        assertEquals(1, 0.numberOfDigits())
```
Comment (soywiz) [inline korlibs-math-core/test/korlibs/math/MathTest.kt:8]: Missing `Int.MAX_VALUE`, `Int.MIN_VALUE` to cover extra edge cases.
We could also add tests for: 1, 10, 100, 1000, 10000..., 9, 99, 999, 9999... 1, 11, 101, 1001, 10001

File: korlibs-math-core/src/korlibs/math/Math.kt (hunk 1 above)
Comment (reosfire) [reply korlibs-math-core/src/korlibs/math/Math.kt:162]: On jvm the results are exactly opposite. Here is the results of jmh benchmark
image

It was written in java:

```
package org.snow.benchmarks;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.infra.Blackhole;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

@State(Scope.Benchmark)
public class SampleBenchmark {

	int[] dataArray = new int[1000000];

	@Setup
	public void setup() {
		Random random = new Random(42);
		for (int i = 0; i < dataArray.length; i++) {
			dataArray[i] = random.nextInt(100);
		}
	}

	@Benchmark
	public void testDivision(Blackhole blackhole) {
		for (int i = 0; i < dataArray.length; i++) {
			blackhole.consume(division(dataArray[i], 10));
		}
	}

	@Benchmark
	public void testLog(Blackhole blackhole) {
		for (int i = 0; i < dataArray.length; i++) {
			blackhole.consume(loging(dataArray[i], 10));
		}
	}

	private int division(int n, int radix) {
		if (n == 0) return 1;
		int digits = 0;
		while (n != 0) {
			n /= radix;
			digits++;
		}
		return digits;
	}

	private int loging(int n, int radix) {
		if (n == 0) return 1;
		return (int) (Math.log(n) / Math.log(radix)) + 1;
	}
}
```

Also I think the results can differ because you are passing floating point number into the functions in your benchmark therefore the division cannot be performed efficiently.

Probably we want somehow use different Implementations per platform. But idk how to do that😅

Comment (soywiz) [reply korlibs-math-core/src/korlibs/math/Math.kt:162]: Adjusting a bit so JS knows parameters are integers like Kotlin would do:

Captura de pantalla 2025-05-05 a las 16 47 14

Only twice as slow, so I guess it is fine, if in JVM is better.

**Comment by soywiz** [comment]
@reosfire tests are failing so it cannot be merged. You have to run `./gradlew apiDump` or jvmApiDump (not sure) so api files are updated. You will also have to add `@JvmOverloads` annotation to `almostEquals` and `almostZero` to keep ABI compatibility.
